sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,251 @@
+	if err != nil {
+		notifyFile.Close()
+		return nil, fmt.Errorf("cannot notify ioctl to modeset user on %q: %v", path, err)
+	}
+
+	poll, err := epoll.Open()
+	if err != nil {
+		notifyFile.Close()
+		return nil, fmt.Errorf("cannot open epoll file descriptor: %v", err)
+	}
+	// XXX: Do we need a notification for Writable, to send responses back?
+	if err := poll.Register(int(notifyFile.Fd()), epoll.Readable); err != nil {
+		notifyFile.Close()
+		poll.Close()
+		return nil, fmt.Errorf("cannot register epoll on %q: %v", path, err)
+	}
+
+	listener := &Listener{
+		R: make(chan *Request),
+		E: make(chan error),
+
+		notifyFile: notifyFile,
+		poll:       poll,
+	}
+	return listener, nil
+}
+
+func (l *Listener) decodeAndDispatchRequest(buf []byte, tomb *tomb.Tomb) error {
+	var nmsg notify.MsgNotification
+	if err := nmsg.UnmarshalBinary(buf); err != nil {
+		return err
+	}
+	// What kind of notification message did we get?
+	switch nmsg.NotificationType {
+	case notify.APPARMOR_NOTIF_OP:
+		var omsg notify.MsgNotificationOp
+		if err := omsg.UnmarshalBinary(buf); err != nil {
+			return err
+		}
+		// What kind of operation notification did we get?
+		switch omsg.Class {
+		case notify.AA_CLASS_FILE:
+			var fmsg notify.MsgNotificationFile
+			if err := fmsg.UnmarshalBinary(buf); err != nil {
+				return err
+			}
+			// log.Printf("notification request: %#v\n", fmsg)
+			req := newRequest(l, &fmsg)
+			l.R <- req
+			tomb.Go(func() error {
+				l.waitAndRespond(req, &fmsg)
+				return nil
+			})
+		default:
+			return fmt.Errorf("unsupported mediation class: %v", omsg.Class)
+		}
+	default:
+		return fmt.Errorf("unsupported notification type: %v", nmsg.NotificationType)
+	}
+	return nil
+}
+
+func (l *Listener) waitAndRespond(req *Request, msg *notify.MsgNotificationFile) {
+	resp := notify.ResponseForRequest(&msg.MsgNotification)
+	resp.MsgNotification.Error = 0 // ignored in responses
+	resp.MsgNotification.NoCache = 1
+	if allow := <-req.YesNo; allow {
+		resp.Allow = msg.Allow | msg.Deny
+		resp.Deny = 0
+		resp.Error = 0
+	} else {
+		resp.Allow = 0
+		resp.Deny = msg.Deny
+		resp.Error = msg.Error
+		// msg.Error is field from MsgNotificationResponse, and is unused.
+		// msg.MsgNotification.Error is also ignored in responses.
+	}
+	//log.Printf("notification response: %#v\n", resp)
+	if err := l.encodeAndSendResponse(&resp); err != nil {
+		l.fail(err)
+	}
+}
+
+func (l *Listener) encodeAndSendResponse(resp *notify.MsgNotificationResponse) error {
+	buf, err := resp.MarshalBinary()
+	if err != nil {
+		return err
+	}
+	ioctlBuf := notify.BytesToIoctlRequestBuffer(buf)
+	_, err = notifyIoctl(l.notifyFile.Fd(), notify.APPARMOR_NOTIF_SEND, ioctlBuf)
+	return err
+}
+
+func (l *Listener) runOnce(tomb *tomb.Tomb) error {
+	// XXX: Wait must return immediately once epoll is closed.
+	events, err := l.poll.Wait()
+	if err != nil {
+		return err
+	}
+	for _, event := range events {
+		switch event.Fd {
+		case int(l.notifyFile.Fd()):
```

Comment:
Reading this I was initially a bit confused why we would need a switch here and the case of l.notifyFile.Fd() and why/if there ever are different cases. Maybe it's just me but to me something like:
```go
@@ -202,21 +202,22 @@ func (l *Listener) runOnce(tomb *tomb.Tomb) error {
                return err
        }
        for _, event := range events {
-               switch event.Fd {
-               case int(l.notifyFile.Fd()):
-                       if event.Readiness&epoll.Readable != 0 {
-                               // Prepare a receive buffer for incoming request. The buffer is of the
-                               // maximum allowed size and will contain one kernel request upon return.
-                               // Note that the actually occupied buffer is indicated by the Length field
-                               // in the header.
-                               ioctlBuf := notify.NewIoctlRequestBuffer()
-                               buf, err := notifyIoctl(l.notifyFile.Fd(), notify.APPARMOR_NOTIF_RECV, ioctlBuf)
-                               if err != nil {
-                                       return err
-                               }
-                               if err := l.decodeAndDispatchRequest(buf, tomb); err != nil {
-                                       return err
-                               }
+               if event.Fd != int(l.notifyFile.Fd()) {
+                       return fmt.Errorf("unexpected event from fd %v (%v)", event.Fd, event.Readiness)
+               }
+
+               if event.Readiness&epoll.Readable != 0 {
+                       // Prepare a receive buffer for incoming request. The buffer is of the
+                       // maximum allowed size and will contain one kernel request upon return.
+                       // Note that the actually occupied buffer is indicated by the Length field
+                       // in the header.
+                       ioctlBuf := notify.NewIoctlRequestBuffer()
+                       buf, err := notifyIoctl(l.notifyFile.Fd(), notify.APPARMOR_NOTIF_RECV, ioctlBuf)
+                       if err != nil {
+                               return err
+                       }
+                       if err := l.decodeAndDispatchRequest(buf, tomb); err != nil {
+                               return err
                        }
                }
        }
```
would feel more naturally (unless I miss something here of course).

Reply:
I agree, the switch seems unnecessary. In case other FDs are registered on the epoll FD as well, I'd prefer a `logger.Debugf` to an error return, and `continue` following that log statement. A `continue` can be used `if event.Readiness&epoll.Readable == 0` as well, making the body even less nested.